From: Nick Piggin <npiggin@suse.de>
Subject: fix prefomance problems with inotify
References: 153850
Patch-mainline: yes

I'm not sure what the cleanest way to do this is. I'm fairly sure the vfs
guys do not want a DENTRY_INOTIFY_ flag in fs/dcache.c, so I've added a
comment there.

I also don't know whether or not the inotify guys want parent events on
negative entries. They don't appear to now, so I'll take that as a no.

After the DENTRY_INOTIFY_PARENT_WATCHED debugging stuff is taken out, that
flag will basically be undefined for negative dentrys rather than always
clear (after this patch). I'm not sure whether the vfs people consider that
to be unclean.

Signed-off-by: Andrew Morton <akpm@osdl.org>
Acked-by: 

---

 fs/dcache.c  |    3 +++
 fs/inotify.c |    5 +++++
 2 files changed, 8 insertions(+)

diff -puN fs/dcache.c~inotify-lock-avoidance-with-parent-watch-status-in-dentry-fix-2 fs/dcache.c
--- devel/fs/dcache.c~inotify-lock-avoidance-with-parent-watch-status-in-dentry-fix-2	2006-03-01 21:41:16.000000000 -0800
+++ devel-akpm/fs/dcache.c	2006-03-01 21:41:16.000000000 -0800
@@ -1177,6 +1177,9 @@ void d_delete(struct dentry * dentry)
 	spin_lock(&dentry->d_lock);
 	isdir = S_ISDIR(dentry->d_inode->i_mode);
 	if (atomic_read(&dentry->d_count) == 1) {
+		/* remove this and other inotify debug checks after 2.6.18 */
+		dentry->d_flags &= ~DCACHE_INOTIFY_PARENT_WATCHED;
+
 		dentry_iput(dentry);
 		fsnotify_nameremove(dentry, isdir);
 		return;
diff -puN fs/inotify.c~inotify-lock-avoidance-with-parent-watch-status-in-dentry-fix-2 fs/inotify.c
--- devel/fs/inotify.c~inotify-lock-avoidance-with-parent-watch-status-in-dentry-fix-2	2006-03-01 21:41:16.000000000 -0800
+++ devel-akpm/fs/inotify.c	2006-03-01 21:41:16.000000000 -0800
@@ -390,6 +390,7 @@ static inline int inotify_inode_watched(
 
 /*
  * Get child dentry flag into synch with parent inode.
+ * Flag should always be clear for negative dentrys.
  */
 static void set_dentry_child_flags(struct inode *inode, int watched)
 {
@@ -400,6 +401,10 @@ static void set_dentry_child_flags(struc
 		struct dentry *child;
 
 		list_for_each_entry(child, &alias->d_subdirs, d_u.d_child) {
+			if (!child->d_inode) {
+				WARN_ON(child->d_flags & DCACHE_INOTIFY_PARENT_WATCHED);
+				continue;
+			}
 			spin_lock(&child->d_lock);
 			if (watched) {
 				WARN_ON(child->d_flags &
_
